src/crlf.c

```diff
@@ -161,9 +161,10 @@ static int crlf_apply_to_odb(
 
 	/* If safecrlf is enabled, sanity-check the result. */
 	if (ca->crlf_action == GIT_CRLF_INPUT ||
-		(ca->auto_crlf == GIT_AUTO_CRLF_INPUT &&
-		(ca->crlf_action == GIT_CRLF_GUESS || ca->crlf_action == GIT_CRLF_AUTO ||
-		(ca->crlf_action == GIT_CRLF_TEXT && ca->eol==GIT_EOL_UNSET)))) {
+		(ca->crlf_action == GIT_CRLF_GUESS || ca->crlf_action == GIT_CRLF_AUTO || ca->crlf_action == GIT_CRLF_TEXT) && ca->eol == GIT_EOL_LF ||
+		(ca->auto_crlf == GIT_AUTO_CRLF_INPUT && (ca->crlf_action == GIT_CRLF_GUESS)) ||
+		(ca->auto_crlf == GIT_AUTO_CRLF_INPUT && (ca->crlf_action == GIT_CRLF_AUTO || ca->crlf_action == GIT_CRLF_TEXT) && ca->eol == GIT_EOL_UNSET) ||
+		(GIT_EOL_NATIVE != GIT_EOL_CRLF && ca->auto_crlf == GIT_AUTO_CRLF_FALSE && (ca->crlf_action == GIT_CRLF_AUTO || ca->crlf_action == GIT_CRLF_TEXT) && ca->eol == GIT_EOL_UNSET)) {
```

Comment:
This seems overcomplicated to me.  For one thing, you sometimes have the action test before the auto_crlf test and sometimes after.  This means that you miss some commonalities like the duplicated ca->crlf_action == auto || ca->crlf_action == text.  We could also make it easier to read by breaking out sa->action, sa->eol, and sa->auto_crlf into separate variables.  And we could define a helper GIT_EOF_NATIVE_IS_LF = GIT_EOL_NATIVE != GIT_EOL_CRLF, just to make that bit a bit shorter and clearer.   I made these changes and I was still unable to make heads or tails of it.  So then I thought: what if we make a function?


```
static int is_crlf_reasonable(sa) {

	switch (sa->crlf_action) {
	case GIT_CRLF_INPUT:
		return 1;
		break;
	case GIT_CRLF_GUESS:
		return sa->eol == GIT_EOL_LF || sa->auto_crlf == GIT_AUTO_CRLF_INPUT;
	case GIT_CRLF_AUTO: /* FALL-THROUGH */
	case GIT_CRLF_TEXT:
		switch (sa->eol) {
		case GIT_EOL_LF:
			return 1;
		case GIT_EOL_UNSET:
			return sa->auto_crlf == GIT_AUTO_CRLF_INPUT;
		case GIT_EOL_UNSET:
		return GIT_EOL_NATIVE_IS_LF && sa->auto_crlf == GIT_AUTO_CRLF_FALSE;
		}
	}

	return 0;
}
```

I have not yet actually tested this to ensure that it is identical, but I think the overall approach is much easier to read. Plus it's clear from the function name that success means that the check passed. Perhaps we need a better function name -- I don't really understand what this code is doing, but I can tell that I would be unable to debug it.